pan3d/viewer.py

```diff
@@ -0,0 +1,265 @@
+import json
+import os
+import pyvista
+from pvxarray.vtk_source import PyVistaXarraySource
+import xarray
+
+from trame.decorators import TrameApp, change
+from trame.ui.vuetify import SinglePageWithDrawerLayout  # TODO: upgrade to vuetify 3
+from trame.app import get_server
+
+from pan3d.ui import initialize as viewer_ui
+from pan3d.utils import initial_state, run_singleton_task
+
+
+@TrameApp()
+class Pan3DViewer:
+    def __init__(self, server=None, dataset_path=None, state=None):
+        if server is None:
+            server = get_server()
+
+        if isinstance(server, str):
+            server = get_server(server)
+
+        self.server = server
+        self.state.update(initial_state)
+        self.algorithm = PyVistaXarraySource()
+        self.plotter = pyvista.Plotter(off_screen=True, notebook=False)
+        self.plotter.set_background("lightgrey")
+        self.dataset = None
+        self.dataset_path = None
+        self.mesh = None
+        self.actor = None
+
+        self.ctrl.get_plotter = lambda: self.plotter
+        self.ctrl.reset = self.reset
+
+        if dataset_path:
+            self.set_dataset_path(dataset_path=dataset_path)
+        if state:
+            self.state.update(state)
+
+        # Fix version of vue
+        server.client_type = "vue2"  # TODO: upgrade to vue3
+        # Build GUI
+        self.layout = SinglePageWithDrawerLayout(self.server)
+        viewer_ui(self.layout, self.ctrl)
```

Comment:
I don't think the UI needs to be externalized. Also the layout has the server, so you have access to the controller from there as well. But if you have everything defined within the class, you don't need the controller. You can directly by your self.method.